[compiled autograd] log compilation time to perfetto

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140964
* #143175
* #141907

https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmprli4iy/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100
```
[
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886868992655.8
  },
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869130681.0
  },
  {
    "args": {
      "compile_id": "0/0/0"
    },
    "cat": "dynamo_timed",
    "name": "dynamo",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869134350.5
  },
  {
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @yf225

xmfan committed:
[compiled autograd] log compilation time via ChromiumEventLogger

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] log compilation time via ChromiumEventLogger"


Compiled Autograd creates a GraphModule and calls torch.compile on it. So it happens before Dynamo and is a separate dimension from CompileId's frame_id/frame_compile_id. That change is not easy to fully support due to the data pipelines relying on the CompileId's string representation. 

Instead we defer the logging to the Dynamo call that always follows the CA compilation. I don't think there is a case where a CA cache miss can be followed by a Dynamo cache hit.

Example tlparse: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpsB0Wm1/index.html
```
  {
    "args": {
      "compile_id": "1/0"
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1731954155070324.8
  },
  {
    "args": {
      "compile_id": "1/0"
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1731954155148152.5
  },
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

ezyang commented:
deferring to @masnesral

masnesral commented:
@xmfan high-level question: Is the ask here to add this timing only to perfetto or do we want these times to show up in the dynamo_compile scuba table?

xmfan commented:
I would want it to show up in both

xmfan committed:
Update on "[compiled autograd] log compilation time to perfetto"


Compiled Autograd (CA) creates a GraphModule and calls torch.compile on it. So it happens before Dynamo and is a separate dimension from CompileId's frame_id/frame_compile_id. That change is not easy to fully support due to the data pipelines relying on the CompileId's string representation. 

Instead we defer the logging to the Dynamo call that always follows the CA compilation. I don't think there is a case where a CA cache miss can be followed by a Dynamo cache hit.

Example tlparse: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpsB0Wm1/index.html
```
  {
    "args": {
      "compile_id": "1/0"
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1731954155070324.8
  },
  {
    "args": {
      "compile_id": "1/0"
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1731954155148152.5
  },
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] log compilation time to perfetto"


Compiled Autograd (CA) creates a GraphModule and calls torch.compile on it. So it happens before Dynamo and is a separate dimension from CompileId's frame_id/frame_compile_id. That change is not easy to fully support due to the data pipelines relying on the CompileId's string representation. 

Instead we defer the logging to the Dynamo call that always follows the CA compilation. I don't think there is a case where a CA cache miss can be followed by a Dynamo cache hit.

Example tlparse: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpsB0Wm1/index.html
```
  {
    "args": {
      "compile_id": "1/0"
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1731954155070324.8
  },
  {
    "args": {
      "compile_id": "1/0"
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1731954155148152.5
  },
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

masnesral commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -86,6 +92,8 @@ def begin_capture(
         origins: List[List[Tuple[int, str]]],
     ):
+        self.start_time = time.time_ns()
+        self.id = counters["compiled_autograd"]["captures"]
```
Is this guaranteed to exist?

xmfan replied:
yes, defaultdict

masnesral commented on torch/_dynamo/convert_frame.py:
```diff
@@ -866,6 +866,21 @@ def count_args(code: CodeType) -> int:
+        locals.get("self")
+    ):
+        ca_id, ca_start_ns, ca_end_ns = ca_metrics
+        ca_event = f"compiled_autograd_{ca_id}"
```
I don't know of any realy problem it would cause, but curious why you did it this way. So far the event names have been static strings.

xmfan replied:
im moving it to the metadata dict, i just want to be able to refer to which graph corresponded to this compile event

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -3355,6 +3355,29 @@ def wrapper(*args):
+    end_time_ns: int
+
+
+def get_compiled_autograd_metrics(maybe_gm):
```
nit: can you add typehints to these new helpers; I think we've been getting a bit more principled about using them in this particular file at least

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -3355,6 +3355,29 @@ def wrapper(*args):
+        or "compiled_autograd_metrics" not in maybe_gm.meta
+    ):
+        return None
+    return dataclasses.astuple(maybe_gm.meta["compiled_autograd_metrics"])
```
Since you created the new CompiledAutogradMetrics, maybe just return it directly instead of converting to a tuple? Makes the usage at the callsite unambiguous at least.

masnesral reviewed (approved):
some nits for you to consider, but LGTM

xmfan committed:
Update on "[compiled autograd] log compilation time to perfetto"


Compiled Autograd (CA) creates a GraphModule and calls torch.compile on it. So it happens before Dynamo and is a separate dimension from CompileId's frame_id/frame_compile_id. That change is not easy to fully support due to the data pipelines relying on the CompileId's string representation. 

Instead we defer the logging to the Dynamo call that always follows the CA compilation. I don't think there is a case where a CA cache miss can be followed by a Dynamo cache hit.

example tlparse: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpttl7Uo/index.html
```
{
    "args": {
      "compile_id": "0/0",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1732324793720137.2
  },
  {
    "args": {
      "compile_id": "0/0",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1732324793859746.8
  },
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

xmfan commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -368,7 +376,7 @@ def end_capture(self, outputs):
         verbose_log.debug("%s", lazy_graph_code)
         trace_structured(
-            "compiled_autograd_graph",
+            f"compiled_autograd_graph_{self.id}",
```
uhh ignore this line, it's being overwritten in the next PR, will land together

xmfan committed:
Update on "[compiled autograd] log compilation time to perfetto"


Compiled Autograd (CA) creates a GraphModule and calls torch.compile on it. So it happens before Dynamo and is a separate dimension from CompileId's frame_id/frame_compile_id. That change is not easy to fully support due to the data pipelines relying on the CompileId's string representation. 

Instead we defer the logging to the Dynamo call that always follows the CA compilation. I don't think there is a case where a CA cache miss can be followed by a Dynamo cache hit.

example tlparse: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpttl7Uo/index.html
```
{
    "args": {
      "compile_id": "0/0",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1732324793720137.2
  },
  {
    "args": {
      "compile_id": "0/0",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1732324793859746.8
  },
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] log compilation time to perfetto"


https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmprli4iy/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100
```
[
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886868992655.8
  },
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869130681.0
  },
  {
    "args": {
      "compile_id": "0/0/0"
    },
    "cat": "dynamo_timed",
    "name": "dynamo",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869134350.5
  },
  {
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] log compilation time to perfetto"


https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmprli4iy/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100
```
[
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886868992655.8
  },
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869130681.0
  },
  {
    "args": {
      "compile_id": "0/0/0"
    },
    "cat": "dynamo_timed",
    "name": "dynamo",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869134350.5
  },
  {
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] log compilation time to perfetto"


https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmprli4iy/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100
```
[
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886868992655.8
  },
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869130681.0
  },
  {
    "args": {
      "compile_id": "0/0/0"
    },
    "cat": "dynamo_timed",
    "name": "dynamo",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869134350.5
  },
  {
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] log compilation time to perfetto"


https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmprli4iy/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100
```
[
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886868992655.8
  },
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869130681.0
  },
  {
    "args": {
      "compile_id": "0/0/0"
    },
    "cat": "dynamo_timed",
    "name": "dynamo",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869134350.5
  },
  {
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

jamesjwu commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -421,6 +435,13 @@ def runtime_wrapper(compiled_fn, inputs, sizes, scalars, hooks):
             finally:
                 in_compiled_autograd_region = False
 
+        get_chromium_event_logger().log_event_end(
```
Is there a chance the code between the start/end events fails or throws an exception? You might want this in a finally clause, or better yet, to use dynamo_timed directly

xmfan committed:
Update on "[compiled autograd] log compilation time to perfetto"


https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmprli4iy/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100
```
[
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886868992655.8
  },
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869130681.0
  },
  {
    "args": {
      "compile_id": "0/0/0"
    },
    "cat": "dynamo_timed",
    "name": "dynamo",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869134350.5
  },
  {
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] log compilation time to perfetto"


https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmprli4iy/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100
```
[
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886868992655.8
  },
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869130681.0
  },
  {
    "args": {
      "compile_id": "0/0/0"
    },
    "cat": "dynamo_timed",
    "name": "dynamo",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869134350.5
  },
  {
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] log compilation time to perfetto"


https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmprli4iy/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100
```
[
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886868992655.8
  },
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869130681.0
  },
  {
    "args": {
      "compile_id": "0/0/0"
    },
    "cat": "dynamo_timed",
    "name": "dynamo",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869134350.5
  },
  {
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] log compilation time to perfetto"


https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmprli4iy/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100
```
[
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886868992655.8
  },
  {
    "args": {
      "compile_id": "0/-/-",
      "graph_id": 0
    },
    "cat": "dynamo_timed",
    "name": "compiled_autograd",
    "ph": "E",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869130681.0
  },
  {
    "args": {
      "compile_id": "0/0/0"
    },
    "cat": "dynamo_timed",
    "name": "dynamo",
    "ph": "B",
    "pid": 0,
    "tid": 0,
    "ts": 1733886869134350.5
  },
  {
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]
